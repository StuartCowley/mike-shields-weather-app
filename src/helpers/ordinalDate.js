export default function ordinalDate(dayOfMonth) {
  if (dayOfMonth > 3 && dayOfMonth < 21) return `${dayOfMonth}th`;
  switch (dayOfMonth % 10) {
    case 1:
      return `${dayOfMonth}st`;
    case 2:
      return `${dayOfMonth}nd`;
    case 3:
      return `${dayOfMonth}rd`;
    default:
      return `${dayOfMonth}th`;
  }
}
