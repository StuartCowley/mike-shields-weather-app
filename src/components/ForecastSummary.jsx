import React from "react";
import PropTypes from "prop-types";
import WeatherIcon from "react-icons-weather";
import ordinalDate from "../helpers/ordinalDate";

export default function ForecastSummary({
  date,
  description,
  icon,
  temperature,
}) {
  const dateString = new Date(date).toDateString();
  const [weekday, month, dayOfMonth] = dateString.split(" ");
  const formattedDate = `${weekday} ${ordinalDate(+dayOfMonth)} ${month}`;

  return (
    <div className="forecast-summary" data-testid="forecast-summary">
      <h2 className="forecast-summary__date">{formattedDate}</h2>
      <p className="forecast-summary__description">{description}</p>
      <WeatherIcon
        className="forecast-summary__icon"
        data-testid="forecast-summary-icon"
        iconId={icon}
        name="owm"
      />
      <p className="forecast-summary__temperature">{temperature.max}</p>
    </div>
  );
}

ForecastSummary.propTypes = {
  date: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired,
  temperature: PropTypes.shape({
    max: PropTypes.number,
    min: PropTypes.number,
  }).isRequired,
};
