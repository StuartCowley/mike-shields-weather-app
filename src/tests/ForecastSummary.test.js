import React from "react";
import { render } from "@testing-library/react";
import ForecastSummary from "../components/ForecastSummary";
import forecast from "../data/forecast.json";
import ordinalDate from "../helpers/ordinalDate";

const dailyForecast = forecast.forecasts[0];
const { date, description, icon, temperature } = dailyForecast;

const dateString = new Date(date).toDateString();
const [weekday, month, dayOfMonth] = dateString.split(" ");
const formattedDate = `${weekday} ${ordinalDate(Number(dayOfMonth))} ${month}`;

describe("ForecastSummary", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(
      <ForecastSummary
        date={date}
        description={description}
        icon={icon}
        temperature={temperature}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders formatted date, description, icon and temperature", () => {
    const { getByText, getByTestId } = render(
      <ForecastSummary
        data-testid="forecast-summary"
        date={date}
        description={description}
        icon={icon}
        temperature={temperature}
      />
    );

    expect(getByText(formattedDate)).toBeInTheDocument();
    expect(getByText(description)).toBeInTheDocument();
    expect(getByTestId("forecast-summary-icon")).toBeInTheDocument();
    expect(getByText(temperature.max)).toBeInTheDocument();
  });

  it("rendered elements have correct class names", () => {
    const { getByText, getByTestId } = render(
      <ForecastSummary
        data-testid="forecast-summary"
        date={date}
        description={description}
        icon={icon}
        temperature={temperature}
      />
    );

    expect(getByText(formattedDate)).toHaveClass("forecast-summary__date");
    expect(getByText(description)).toHaveClass("forecast-summary__description");
    expect(getByTestId("forecast-summary-icon")).toHaveClass(
      "forecast-summary__icon"
    );
    expect(getByText(temperature.max)).toHaveClass(
      "forecast-summary__temperature"
    );
  });
});
