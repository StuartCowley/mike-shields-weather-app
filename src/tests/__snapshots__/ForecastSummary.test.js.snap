// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastSummary matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-summary"
    data-testid="forecast-summary"
  >
    <h2
      class="forecast-summary__date"
    >
      Mon 30th Apr
    </h2>
    <p
      class="forecast-summary__description"
    >
      Clear
    </p>
    <i
      class="wi wi-day-sunny forecast-summary__icon"
      data-testid="forecast-summary-icon"
      name="owm"
    />
    <p
      class="forecast-summary__temperature"
    >
      11
    </p>
  </div>
</DocumentFragment>
`;
